src/custom.rs

```diff
@@ -0,0 +1,31 @@
+// Copyright 2018 Developers of the Rand project.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! An implementation which calls out to an externally defined function.
+use crate::Error;
+
+/// Reister a function to be invoked by getrandom on custom targets. This macro
+/// is only avalible on custom targets to prevent registering a function that
+/// will not be called.
+#[macro_export]
+macro_rules! register_custom_getrandom {
+    ($path:path) => {
+        #[no_mangle]
+        fn __getrandom_custom(dest: &mut [u8]) -> Result<(), Error> {
+            $path(dest)
+        }
+    };
+}
+
+pub fn getrandom_inner(dest: &mut [u8]) -> Result<(), Error> {
+    extern "Rust" {
+        #[allow(improper_ctypes)] // See rust-lang/rust#64593
+        fn __getrandom_custom(dest: &mut [u8]) -> Result<(), Error>;
```

Comment:
Looks like a nice solution.

The call stack is now potentially `getrandom::getrandom` → `getrandom::getrandom_inner` → `__getrandom_custom` → `provider::getrandom_inner`, but use of `#[inline]` in a couple of places reduces this to 2.

Reply:
I think even without the `#[inline]` Rust will manage to get rid of all the intermediate function calls except the `__getrandom_custom`, but I should run a compile test to make sure.

Reply:
Yes, and I think `__getrandom_custom` can never be inlined, but it shouldn't be a big issue.